### PR TITLE
ipq40xx: Add support for Plasma Cloud PA1200 and PA2200

### PR DESCRIPTION
**Pull request by ecsv**
These changes depend on ~~#3614~~ and #3616 (this is why you see them here in the list of commits).

Plasma Cloud PA1200
===================

* QCA IPQ4018
* 256 MB of RAM
* 32 MB of SPI NOR flash (w25q256)
  - 2x 15 MB available; but one of the 15 MB regions is the recovery image
* 2T2R 2.4 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=PlasmaCloud-PA1200
* 2T2R 5 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=PlasmaCloud-PA1200
* 3x GPIO-LEDs for status (cyan, purple, yellow)
* 1x GPIO-button (reset)
* 1x USB (xHCI)
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 2x gigabit ethernet
  - phy@mdio4:
    + Label: Ethernet 1
    + gmac0 (ethaddr) in original firmware
    + used as LAN interface
  - phy@mdio3:
    + Label: Ethernet 2
    + gmac1 (eth1addr) in original firmware
    + 802.3af/at POE(+)
    + used as WAN interface
* 12V/24V 1A DC

Plasma Cloud PA2200
===================

* QCA IPQ4019
* 256 MB of RAM
* 32 MB of SPI NOR flash (w25q256)
  - 2x 15 MB available; but one of the 15 MB regions is the recovery image
* 2T2R 2.4 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=20,variant=PlasmaCloud-PA2200
* 2T2R 5 GHz (channel 36-64)
  - QCA9888 hw2.0 (PCI)
  - requires special BDF in QCA9888/hw2.0/board-2.bin
    bus=pci,bmi-chip-id=0,bmi-board-id=16,variant=PlasmaCloud-PA2200
* 2T2R 5 GHz (channel 100-165)
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=21,variant=PlasmaCloud-PA2200
* GPIO-LEDs for 2.4GHz, 5GHz-SoC and 5GHz-PCIE
* GPIO-LEDs for power (orange) and status (blue)
* 1x GPIO-button (reset)
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 2x gigabit ethernet
  - phy@mdio3:
    + Label: Ethernet 1
    + gmac0 (ethaddr) in original firmware
    + used as LAN interface
  - phy@mdio4:
    + Label: Ethernet 2
    + gmac1 (eth1addr) in original firmware
    + 802.3at POE+
    + used as WAN interface
* 12V 2A DC

Flashing
========

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
used to transfer the factory image to the u-boot when the device boots up.

### CODE REVIEW

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts:17]: Please use an SPDX license identifier instead.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+		led_cyan: cyan {
+			label = "cyan:status";
+			gpios = <&tlmm 0 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_purple: purple {
+			label = "purple:status";
+			gpios = <&tlmm 1 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "default-off";
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts:116]: This line can be dropped, it's default.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	status = "okay";
+};
+
+&blsp1_spi1 {
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
+
+	m25p80@0 {
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts:164]: flash@0 ?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts:17]: SPDX ...

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+		led-boot = &led_purple;
+		led-failsafe = &led_yellow;
+		led-running = &led_cyan;
+		led-upgrade = &led_yellow;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_cyan: cyan {
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-pa1200.dts:108]: `led_status_cyan: status_cyan {`

and so on ...

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+		led-boot = &led_power;
+		led-failsafe = &led_blue;
+		led-running = &led_power;
+		led-upgrade = &led_blue;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_power: orange {
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts:99]: led_power_orange: power_orange {

...

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+	status = "okay";
+};
+
+&blsp1_spi1 {
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+	cs-gpios = <&tlmm 12 GPIO_ACTIVE_HIGH>;
+
+	m25p80@0 {
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts:165]: flash@0

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <24000000>;
+		/* partitions are passed via bootloader */
+	};
+};
+
+
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-pa2200.dts:175]: drop second empty line

File: target/linux/ipq40xx/image/Makefile [hunk 9]
```diff
@@ -666,6 +666,23 @@ define Device/plasmacloud_pa1200
 endef
 TARGET_DEVICES += plasmacloud_pa1200
 
+
```
Comment (adschm) [inline target/linux/ipq40xx/image/Makefile:669]: don't add second empty line.

**Comment by ynezz** [comment]
Merged into my staging tree, thanks

**Pull request closed without merging**